### ISSUE TITLE
`BasePwCpInputGenerator`: Improve `parameters` input validation

### ISSUE THREAD

**Issue by mbercx** [aiidateam/aiida-quantumespresso#675]
### Problem description

Currently, the `parameters` input for the `BasePwCpInputGenerator` is:

1. Not validated as a validator, but in the `prepare_for_submission`  method, raising `InputValidationError`s when there are issues.
2. Lacking a lot of validation that causes e.g. `pw.x` to crash.

[1] Causes the calculation to `Except`, e.g. when using the `PwBaseWorkChain.get_builder_from_protocol()` method:

```Python
PwBaseWorkChain = WorkflowFactory('quantumespresso.pw.base')

builder = PwBaseWorkChain.get_builder_from_protocol(
    structure=orm.load_node(3),
    code=orm.load_code(1)
)
builder.pw.parameters['CONTROL']['calculation'] = 'GME stock'
```
submitting this `builder` leads to an `Excepted` work chain with the following error:

```Python
aiida.common.exceptions.InputValidationError: Unknown 'calculation' value in CONTROL namelist GME stock. Otherwise, specify the list of namelists using the NAMELISTS inside the 'settings' input node
```

[2] Usually starts running just fine, but `pw.x` will crash, leading to `exit_status` `305` for the work chain, e.g.:

```Python
PwBaseWorkChain = WorkflowFactory('quantumespresso.pw.base')

builder = PwBaseWorkChain.get_builder_from_protocol(
    structure=orm.load_node(3),
    code=orm.load_code(1)
)
builder.pw.parameters['CONTROL']['wf_collect'] = 'at the moon'
```
submitting this `builder` leads to `exit_status` `305`, which requires the user to use `verdi calcjob outputcat` to figure out what's going wrong by looking at the outputs:

```console
 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
     Error in routine  read_namelists (1):
      bad line in namelist &control: "/" (error could be in the previous line)
 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
```
and even then the user still has to look at the input files to figure out what went wrong.

### Suggested improvements

Use validation based on `jsonschema` along with a `validate_parameters` static method assigned as a validator to the `parameters` input. Move validation that cannot be validated with the schema from the `prepare_for_submission` or `_generate_PWCPinputdata` methods to the `validate_parameters` method, and remove validation that has become unnecessary with the `jsonschema` validation.

### Open questions

- [ ] Should the `validator` be allowed to change the input? An example here is the following fix executed in the `_generate_PWCPinputdata` method:

  ```Python
          input_params = _uppercase_dict(parameters.get_dict(), dict_name='parameters')
          input_params = {k: _lowercase_dict(v, dict_name=k) for k, v in input_params.items()}
  ```
  I'd be inclined to move this to the `validate_parameters` method.

- [ ] Currently, the only `jsonschema` validation we have is for the `PdosWorkChain`. Here, the `jsonschema` is stored in the module itself, defined as a dictionary. I'd be inclined to define the schemas in `yaml` format and place them all in a `schemas` folder as a subfolder of the `calculations` folder, with the following structure:
  ```console
  └── schemas
      ├── parameters.yaml
      └── pw
          └── parameters.yaml
      └── cp
          └── parameters.yaml
  ...
  ```
  Where the top level `parameters.yaml` file contains parameters for the `BasePwCpInputGenerator` class, i.e. that are shared by the `pw.x` and `cp.x` calculations.

- [ ] Are there such schemas already available for Quantum ESPRESSO? Although it would not be an insurmountable task to set these up ourselves, if each released version of QE would have schemas for the inputs, that'd make life quite a bit easier. 😅 
- [ ] How to make the validation compatible with multiple QE versions, and how far back this support should extend.

### Other context

There is quite a bit of validation in the `calculation.helpers.pw_input_helper()` method. A couple of points here:

* If we make the `validate_parameters` a `staticmethod` of the `BasePwCpInputGenerator` and `PwCalculation` classes, the user can validate the inputs parameters via (following the above examples):

  ```Python
  PwCalculation.validate_parameters(builder.pw.parameters, _)
  ```

  For example [1], this would lead to:

  ```Python
  ValidationError: 'GME stock' is not one of ['scf', 'nscf', 'bands', 'relax', 'md', 'vc-relax', 'vc-md']

  Failed validating 'enum' in schema['properties']['CONTROL']['properties']['calculation']:
      {'default': 'scf',
       'enum': ['scf', 'nscf', 'bands', 'relax', 'md', 'vc-relax', 'vc-md'],
       'type': 'string'}

  On instance['CONTROL']['calculation']:
      'GME stock'
  ```
  This would be easier to find than the `pw_input_helper` method.
* Since the `pw_input_helper` method also considers the `structure` input, not all validation that is done here can be moved to the `validate_parameters` method.

**Comment by mbercx** [comment]
Looking at the `pw_input_helper` method some more, it already uses the XML schemas provided and supports different versions (although the latest version have not been added). Using this method for the input validation seems like a decent path forward, then? I suppose one issue is that the QE `version` is not know when submitting the calculation. 🤔

**Comment by sphuber** [comment]
I am all for more validation, all though we cannot guarantee nor should we advertise that we do everything. But already catching common errors would be useful. Using the XML schemas for this would be ideal if possible. The version of QE will indeed be not necessarily known, but we could potentially get this from the `Code` input. Maybe we can start a convention where people will set an extra (or even attribute if this become sort of a standard) that stores the code version. But in that case we won't be able to rely on it, but if there we can provide extra functionality.

> Where are these from? I can find the https://github.com/QEF/qeschemas repo, but this only seems to contain the output .xml schemas?

I am not a 100% sure, but I think they might be simply in the QE source code. I think they are in the same folder as the docs as the xml definitions are also used to generate the docs.

Then regarding changing inputs in the validator: you can't and probably shouldn't. The validators are not made to be changing inputs and in fact I don't think you can. If you were to make changes, they won't be available when you call `self.inputs` within a work chain step.

**Comment by mbercx** [comment]
> I am all for more validation, all though we cannot guarantee nor should we advertise that we do everything.

Yeah, this issue was actually born from going down a validation rabbit hole when I was looking into #661. 😅 Besides improving the validation, I also wanted to separate the input validation from the generation of the input file, and clean up the `_generate_PWCPinputdata` method a bit.

> Maybe we can start a convention where people will set an extra (or even attribute if this become sort of a standard) that stores the code version. But in that case we won't be able to rely on it, but if there we can provide extra functionality.

Seems reasonable. I was also considering looking into a way of comparing e.g. all v6.X schemas and checking if we can produce a reduced version that accept inputs that haven't changed in minor updates of v6.X. That way we can use this as a default. In case the validation fails, we can also print a message on specifying the version of QE in the code extras.

Unfortunately, upon closer inspection, the files in the `calculations/helpers` folder aren't actually XML schemas (`.xsd`) but just XML files, indeed generated for the Docs (although not anymore, it seems, looking at the [`Makefile` for the pw.x documentation](https://gitlab.com/QEF/q-e/-/blob/develop/PW/Doc/Makefile); this could explain why we only have up to v6.4 in the `calculations/helpers` folder). I'm wondering if we can't write a script to convert these into JSON schemas for the input `parameters`. I think it should be fairly easy to then also look for the overlap of these JSON schemas as discussed in the previous paragraph.

> Then regarding changing inputs in the validator: you can't and probably shouldn't.

Indeed, a bit of field testing indicates that any mutation of the `value` is not persistent. I suppose the value is copied at some point in the validation? That does mean that in case we add validation of the `parameters`, using e.g. `'control'` will no longer be fixed automatically (as is now done in the `_generate_PWCPinputdata` method), but fail when e.g. the input is passed to the builder. But perhaps that's a good thing.
